src/measure_theory/integral/layercake.lean

```diff
@@ -0,0 +1,231 @@
+/-
+Copyright (c) 2022 Kalle Kytölä. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kalle Kytölä
+-/
+import measure_theory.integral.interval_integral
+import analysis.special_functions.integrals
+
+/-!
+# The layer cake formula, a.k.a. Cavalieri's principle
+
+In this file we prove the layer cake formula using Fubini's theorem and the two most common
+applications of it: a representation of the integral of a nonnegative function and a representation
+of the integral of the p:th power of a nonnegative function in terms of the "complementary
+cumulative distribution functions" associated with to the measure and the function.
+
+## Main definitions
+
+No new definitions.
+
+## Main results
+
+* `layercake`
+
+## Tags
+
+layer cake representation, Cavalieri's principle
+-/
+
+noncomputable theory
+open_locale ennreal measure_theory
+open set measure_theory filter
+
+/-! ### Layercake theorem -/
+section layercake
+
+namespace measure_theory
+
+variables {α : Type*} [measurable_space α] {f : α → ℝ} {g : ℝ → ℝ} {s : set α}
+
+/-- An auxiliary version of the layer cake theorem (Cavalieri's principle), with a
+measurability assumption that would also essentially follow from the integrability assumptions.
+
+See `measure_theory.layercake` for the main formulation of the layer cake theorem. -/
+lemma layercake_of_measurable (μ : measure α) [sigma_finite μ]
+  (f_nn : 0 ≤ f) (f_mble : measurable f)
+  (g_intble : ∀ t > 0, interval_integrable g volume 0 t)
```

Comment:
This condition is equivalent to `∀ t ≥ 0, interval_integrable g volume 0 t`, which might be easier to use. Why did you choose the strict inequality?

Reply:
Also, would it be better to switch to `∀ t, 0 \le t \to interval_integrable g volume 0 t`?

Reply:
My reasons were totally superficial! Requiring `∀ t ≥ 0, ...` would have been an a priori stronger assumptions than `∀ t > 0, ...`, making the statement "weaker" (of course not really, since the extra requirement is trivial!). I guess I thought such is mathlib style, based on for example preferring `x ≠ ∞` to `x < ∞` in hypotheses about ennreals, and the converse preference in conclusions. But I'm happy to favor convenience of use between these two.

I can also see that Jason's phrasing would put the inequality in the more canonical direction, but I think the presence of the extra (explicit) implication makes the phrasing a bit less readable to mathematicians, which I'm not a fan of.

Reply:
Given that no one (me included) seems to recognize the name "layer cake" for this common result, also the naming here should clearly be improved.

My first attempt might be
 * `layercake` -> `lintegral_primitive_comp_eq_lintegral_measure_tail_mul` for the general case
 * `layercake_one` -> `lintegral_eq_lintegral_measure_tail` for the most standard case
 * `layercake_p` -> `lintegral_rpow_comp_eq_lintegral_measure_tail_mul_deriv_rpow` for the p:th power case.

None of these seem very succinct, and the last one is pretty horrible already (in addition to being misleading about the presence of a literal derivative). In real life, I probably refer to this as "the Fubini trick with complementary cumulative distribution functions", but that doesn't lend itself to appropriate naming in mathlib.

With inspiration from Jason's PR, the maybe the p:th power case could be named `lintegral_rpow_comp_eq_lintegral_measure_le`? (Since `measure_le` seems more mathliby than `measure_tail`, that change should probably be applied to all of the suggestions above.)

Does anyone have good ideas? Should we have a poll on Zulip about the naming?

Reply:
I think since this is a very specific lemma it will be very hard to guess the name if someone is not already familiar with it. So in that case I'd prefer a shorter name

Reply:
> This condition is equivalent to `∀ t ≥ 0, interval_integrable g volume 0 t`, which might be easier to use. Why did you choose the strict inequality?

I thought about this again.

Moving from the assumption `∀ t > 0, ...` to the `∀ t ≥ 0, ...` is currently 5 lines of code in the proof of the auxiliary result `layercake_of_measurable`. I of course agree these lines are rather silly. But I think it would be strictly worse if the users would have the "weaker" assumption in their context and had to reproduce those five silly lines to apply this result. The difference is where to cover the case of integrability on an empty set, and I think it is better to do it once within the theorem rather than to risk leaving that verification to users. I of course might be wrong, and users will always naturally end up with the "stronger" form of the assumption readily available. But not being confident enough about this, I might mildly prefer leaving the statement and proof in their current form.

Reply:
That gt vs ge question is not very important anyway. The way you have it follows our convention of putting "weaker" hypotheses in lemmas in those cases where several are equivalent, so let's keep it like that.